### ISSUE TITLE
usage of write-file-atomic

### ISSUE THREAD

**Issue by mcollina** [ipfs/js-ipfs-repo#183]
Writing file atomically is quite important for lock files and things things like that, however it should hardly be needed for all files operation. In the datastore-fs implementation, all files are written using the [`write-file-atomic`](https://github.com/npm/write-file-atomic) module. As you can see in its source code, it does quite a few file system operations.

See https://github.com/ipfs/js-datastore-fs/blob/a6fc16ab4bd4c0b5d4189591057ae45e0d239908/src/index.js#L13 where write-file-atomic is used.

I think the best goal here is to add an option to `store.put()` to note that if file should be locked or not before usage. The flag could be ignored in other storage systems.

**Comment by daviddias** [comment]
> Writing file atomically is quite important for lock files and things things like that, however it should hardly be needed for all files operation

Agreed. I believe this write-file-atomic got in for all operations because users were reporting doing multiple sequential ipfs.add and ipfs.get and although the first one succeeded, the second one failed because the files didn't have the chance to get flushed to disk.

An alternative to this that will also improve perf is to have a warm in memory cache of all the recent writes so that nextTick reads hit the cache rather then having to go all the way down to the disk.

**Comment by achingbrain** [comment]
> users were reporting doing multiple sequential ipfs.add and ipfs.get and although the first one succeeded, the second one failed because the files didn't have the chance to get flushed to disk

That's interesting - it'd be good to see if this is still a problem with the latest codebase without `write-file-atomic`.  Are there some old issues somewhere about this?

**Comment by daviddias** [comment]
From IRC

<img width="907" alt="image" src="https://user-images.githubusercontent.com/1211152/64522261-460a8280-d302-11e9-831a-1451e97ab551.png">
<img width="848" alt="image" src="https://user-images.githubusercontent.com/1211152/64522275-4efb5400-d302-11e9-9e87-487c6aea2e5f.png">

**Comment by daviddias** [comment]
We need:
- Have a repo config/policy that lets the user specify (with sane defaults)
  - To guarantee that every write is flushed
  - To have an in memory cache for faster confirmation times/reads
- Agree on what are the sane defaults (might depend by runtime and OS, e.g. difference between two OS or OS vs Browser)
